File: scripts/telegram_notifier.js
import fs from "fs";
import https from "https";

const TELEGRAM_BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN;
const TELEGRAM_CHAT_ID = process.env.TELEGRAM_CHAT_ID;

function sendTelegramVideo(videoPath, caption) {
  return new Promise((resolve, reject) => {
    if (!TELEGRAM_BOT_TOKEN || !TELEGRAM_CHAT_ID) {
      console.error("ERROR: Telegram bot token or chat ID not set");
      reject(new Error("Missing credentials"));
      return;
    }

    if (!fs.existsSync(videoPath)) {
      console.error(`ERROR: Video file not found: ${videoPath}`);
      reject(new Error("Video file not found"));
      return;
    }

    const boundary = "----WebKitFormBoundary" + Date.now();
    const videoData = fs.readFileSync(videoPath);

    // Build multipart form data with proper line endings
    const preVideo = [
      `--${boundary}`,
      `Content-Disposition: form-data; name="chat_id"`,
      "",
      TELEGRAM_CHAT_ID,
      `--${boundary}`,
      `Content-Disposition: form-data; name="caption"`,
      "",
      caption,
      `--${boundary}`,
      `Content-Disposition: form-data; name="parse_mode"`,
      "",
      "HTML",
      `--${boundary}`,
      `Content-Disposition: form-data; name="video"; filename="classic_video.mp4"`,
      `Content-Type: video/mp4`,
      ""
    ].join("\r\n");

    const postVideo = `\r\n--${boundary}--`;

    const totalLength = Buffer.byteLength(preVideo) + videoData.length + Buffer.byteLength(postVideo);

    const options = {
      hostname: "api.telegram.org",
      path: `/bot${TELEGRAM_BOT_TOKEN}/sendVideo`,
      method: "POST",
      headers: {
        "Content-Type": `multipart/form-data; boundary=${boundary}`,
        "Content-Length": totalLength
      }
    };

    const req = https.request(options, (res) => {
      let body = "";
      res.on("data", (chunk) => body += chunk);
      res.on("end", () => {
        try {
          const response = JSON.parse(body);
          if (response.ok) {
            resolve(response);
          } else {
            reject(new Error(response.description));
          }
        } catch (e) {
          reject(e);
        }
      });
    });

    req.on("error", reject);
    
    // Write data in chunks to avoid memory issues
    req.write(preVideo);
    req.write(videoData);
    req.write(postVideo);
    req.end();
  });
}

async function main() {
  const videoPath = "./output/video.mp4";
  const topic = process.env.TOPIC || "random classic literature";

  console.log("Starting Telegram notification...");
  console.log(`Video path: ${videoPath}`);
  console.log(`Topic: ${topic}`);

  if (!fs.existsSync(videoPath)) {
    console.error(`ERROR: Video file not found: ${videoPath}`);
    process.exit(1);
  }

  console.log("Video file exists, checking size...");
  const videoStats = fs.statSync(videoPath);
  console.log(`Video size: ${videoStats.size} bytes`);

  // Read social media descriptions
  let tiktokCaption = "No TikTok description available";
  let instagramCaption = "No Instagram description available";
  let hashtags = "";

  console.log("Reading social media files...");
  try {
    if (fs.existsSync("./output/tiktok.txt")) {
      tiktokCaption = fs.readFileSync("./output/tiktok.txt", "utf-8").trim();
      console.log(`TikTok caption loaded (${tiktokCaption.length} chars)`);
    } else {
      console.log("TikTok file not found");
    }
    if (fs.existsSync("./output/instagram.txt")) {
      instagramCaption = fs.readFileSync("./output/instagram.txt", "utf-8").trim();
      console.log(`Instagram caption loaded (${instagramCaption.length} chars)`);
    } else {
      console.log("Instagram file not found");
    }
    if (fs.existsSync("./output/hashtags.txt")) {
      hashtags = fs.readFileSync("./output/hashtags.txt", "utf-8").trim();
      console.log(`Hashtags loaded (${hashtags.length} chars)`);
    } else {
      console.log("Hashtags file not found");
    }
  } catch (error) {
    console.warn("Warning: Could not read social media files:", error.message);
  }

  const caption = `<b>🎬 New Russian Classics Video</b>\n\n<b>Topic:</b> ${topic}\n\n<b>TikTok:</b>\n${tiktokCaption}\n\n<b>Instagram:</b>\n${instagramCaption}\n\n<b>Hashtags:</b>\n${hashtags}`;

  console.log("Sending video to Telegram...");
  console.log(`Caption length: ${caption.length} chars`);
  try {
    await sendTelegramVideo(videoPath, caption);
    console.log("✅ Video sent successfully!");
  } catch (error) {
    console.error("❌ Failed to send video:", error.message);
    console.error("Error details:", error);
    process.exit(1);
  }
}

main().catch((err) => {
  console.error("❌ Error:", err.message);
  process.exit(1);
});

File: telegram_notifier.py
#!/usr/bin/env python3
"""
Telegram bot for sending video and descriptions to user
Sends generated video, TikTok description, Instagram description, and hashtags
"""

import os
import requests
from pathlib import Path

TELEGRAM_BOT_TOKEN = os.environ.get("CLASSIC_TELEGRAM_BOT_TOKEN")
TELEGRAM_CHAT_ID = os.environ.get("TELEGRAM_CHAT_ID")
TOPIC = os.environ.get("TOPIC", "random classic literature")


def send_telegram_message(text: str) -> bool:
    """
    Send text message via Telegram bot
    
    Args:
        text: Message text
    
    Returns:
        True if successful, False otherwise
    """
    if not TELEGRAM_BOT_TOKEN or not TELEGRAM_CHAT_ID:
        print("ERROR: Telegram bot token or chat ID not set")
        return False
    
    url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage"
    payload = {
        "chat_id": TELEGRAM_CHAT_ID,
        "text": text,
        "parse_mode": "HTML"
    }
    
    try:
        response = requests.post(url, json=payload)
        response.raise_for_status()
        return True
    except requests.exceptions.RequestException as e:
        print(f"ERROR: Failed to send Telegram message: {e}")
        return False


def send_telegram_video(video_path: str, caption: str) -> bool:
    """
    Send video via Telegram bot
    
    Args:
        video_path: Path to video file
        caption: Video caption
    
    Returns:
        True if successful, False otherwise
    """
    if not TELEGRAM_BOT_TOKEN or not TELEGRAM_CHAT_ID:
        print("ERROR: Telegram bot token or chat ID not set")
        return False
    
    url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendVideo"
    
    try:
        with open(video_path, 'rb') as video_file:
            payload = {
                "chat_id": TELEGRAM_CHAT_ID,
                "caption": caption,
                "parse_mode": "HTML"
            }
            files = {
                "video": video_file
            }
            response = requests.post(url, data=payload, files=files)
            response.raise_for_status()
            return True
    except requests.exceptions.RequestException as e:
        print(f"ERROR: Failed to send Telegram video: {e}")
        return False


if __name__ == "__main__":
    output_dir = Path("output")
    
    if not output_dir.exists():
        print("ERROR: output directory not found")
        exit(1)
    
    # Read TikTok description
    tiktok_file = output_dir / "tiktok.txt"
    if tiktok_file.exists():
        tiktok_caption = tiktok_file.read_text(encoding="utf-8").strip()
        print(f"TikTok caption loaded ({len(tiktok_caption)} chars)")
    else:
        print("TikTok file not found, using default")
        tiktok_caption = "No TikTok description available"
    
    # Read Instagram description
    instagram_file = output_dir / "instagram.txt"
    if instagram_file.exists():
        instagram_caption = instagram_file.read_text(encoding="utf-8").strip()
        print(f"Instagram caption loaded ({len(instagram_caption)} chars)")
    else:
        print("Instagram file not found, using default")
        instagram_caption = "No Instagram description available"
    
    # Read hashtags
    hashtags_file = output_dir / "hashtags.txt"
    if hashtags_file.exists():
        hashtags = hashtags_file.read_text(encoding="utf-8").strip()
        print(f"Hashtags loaded ({len(hashtags)} chars)")
    else:
        print("Hashtags file not found, using default")
        hashtags = ""
    
    # Video path
    video_path = output_dir / "video.mp4"
    
    if not video_path.exists():
        print(f"ERROR: Video file not found: {video_path}")
        exit(1)
    
    print(f"Video size: {video_path.stat().st_size} bytes")
    
    # Send video
    print("Sending video to Telegram...")
    video_caption = f"<b>🎬 New Russian Classics Video</b>\n\n<b>Topic:</b> {TOPIC}\n\n<b>TikTok:</b>\n{tiktok_caption}\n\n<b>Instagram:</b>\n{instagram_caption}\n\n<b>Hashtags:</b>\n{hashtags}"
    
    if send_telegram_video(str(video_path), video_caption):
        print("✅ Video sent successfully!")
    else:
        print("❌ Failed to send video")
        exit(1)
